dev/com.ibm.ws.jpa.hybridpersistenceactivator/src/com/ibm/ws/javaee/persistence/internal/HybridPersistenceActivator.java

```diff
@@ -23,14 +23,16 @@
 import org.apache.geronimo.osgi.locator.ProviderLocator;
 import org.apache.geronimo.specs.jpa.PersistenceActivator;
 
+import com.ibm.ws.runtime.metadata.ModuleMetaData;
+import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
+
 /**
  * An override of the default persistence activator that provides a hybrid mechanism
  * to discover persistence providers by searching for providers available from
  * META-INF/services in addition to those registered in the OSGi service registry.
  */
 public class HybridPersistenceActivator extends PersistenceActivator {
-
-    private volatile WeakHashMap<ClassLoader, List<PersistenceProvider>> providerCache = new WeakHashMap<ClassLoader, List<PersistenceProvider>>();
+    private volatile WeakHashMap<ModuleMetaData, List<PersistenceProvider>> providerCache = new WeakHashMap<ModuleMetaData, List<PersistenceProvider>>();
```

Comment:
WeakHashMap is not thread safe.  Nothing is done to give you thread safety when accessing the map.  This can lead to infinite loops.

Reply:
Protected map access with a sync block.

Reply:
@jhanders34 please re-review.

Reply:
Yeah it is going to need some work still.  You are going to have to do like the previous code to do the put after you populate the List.  If you don't you can erroneously return an empty or partial list in another thread.

Reply:
Believe this is addressed now.